### ISSUE TITLE
Why does the reserve method search for jobs locked by the worker?

### ISSUE THREAD

**Issue by contentfree** [collectiveidea/delayed_job_mongoid#46]
This check makes the query & sort more inefficient (by using both locked_by and locked_at in an $or) and then sorting to bring the locked_by job to the top. 

Doing so brings mongodb to a screeching halt (100% locked db with just a handful of workers, all of them locking the db to do a findAndModify). It gets really bad when you have even a tiny bit of load (10-20k jobs).

If possible, moving the "Am I already working on something" check outside of the findAndModify would probably speed things up.

**Comment by dblock** [comment]
This is old. Doing some bookkeeping. This looks like a decent suggestion, maybe you want to check this out again @contentfree?

**Comment by johnnyshields** [comment]
Since 2013 when this was raise, MongoDB added support for multiple indexes in OR queries, and we're now indexing this properly. This can be closed. I did find one optimization for which we can make a PR.

```
              { locked_at: nil },
              { locked_at: { '$lt' => (right_now - max_run_time) } }
```

can be combined to:

```
              { locked_at: { '$not' => { '$gte' => (right_now - max_run_time) } } }
```

I will test this on real data if the latter is indeed more efficient.

**Comment by johnnyshields** [comment]
Fixed this in #90
